### ISSUE TITLE
Feature request: Add 'Json' export type support (Export pester object -Passthru).

### ISSUE THREAD

**Issue by Stephanevg** [pester/Pester#1322]
```
I think we have discuss this eventuallity at PsConfEU early june, so here is the issue ;)
```

It would be great to have a way to export the PesterObject in a standard way to pesist it to a file.
The NunitXml is great, but it is not as convenient to parse than the PesterObject.

I can imagine that a ```json``` value could be added to the ```-OutputFormat``` switch

I might be simplifying things, but it could be as easy as simply piping the $PesterObject to ConvertTo-Json before sending it back.
I would be ready to do it, might need some guidance on the where's etc..
Does this make sense?

**Comment by renehernandez** [comment]
@Stephanevg Could you elaborate a bit more on what would be the use case of the json file? Would you want to use it as part of a CI process?

Or as storage mechanisms for the results?

**Comment by Stephanevg** [comment]
Well, potentially it could.
The idea behind this, is that the Pester object that is returned using the ```-Pasthru``` parameter, is a a very convenient way to parse through the test results.

For example, if you want to `presist` your tests for further use (Here I save the results locally, but image this is on a share, or uploaded into a DB perhaps)

Doing this is easier to manipulate **afterwards**, once you need to get back to your objet, rather than parsing the XML document.

```
 $Object = invoke-pester C:\Users\taavast3\OneDrive\Repo\Projects\OpenSource\MeasureScript\Tests -PassThru
 $object | ConvertTo-Json | out-File -FilePath .\TestREsults.json
```

Rather Than:

```
 invoke-pester C.\MyFolderWith\Tests\ -OutputFile .\TestResults.xml
#Parsing the XML file
```

Json is also more lightweight.

Is see this more in Infra testing scenarios, where I want to compare two same tests for example, which happend on a different time.

To sum up: It would be great, if we can have a feature, which allows us to **persist the pester object on disk using in a standard file format**. (I was asking for JSON, as it is prettty much the most common format out there for the moment.

**Comment by Stephanevg** [comment]
@nohwnd Is that something that could be implemented you think?

**Comment by nohwnd** [comment]
I am not sure about this. The xml export is built into Pester because it is complicated, and for historical reason it is configured via Invoke-Pester rather than by piping the output into a cmdlet. This is imho not the best design and I would like to avoid overloading Invoke-Pester with so many features in v5. 

So personally the "workaround" where you pipe output of Pester into ConvertTo-Json is a better approach to me and less complicated than adding it into Invoke-Pester.

**Comment by nohwnd** [comment]
Looking at the PR it seems like a simple change so why not. :)

**Comment by Stephanevg** [comment]
@nohwnd I am bit confused with your two comments which seem to go once against each other.  Is this one still relevant? 


> I am not sure about this. The xml export is built into Pester because it is complicated, and for historical reason it is configured via Invoke-Pester rather than by piping the output into a cmdlet. This is imho not the best design and I would like to avoid overloading Invoke-Pester with so many features in v5.
> 
> So personally the "workaround" where you pipe output of Pester into ConvertTo-Json is a better approach to me and less complicated than adding it into Invoke-Pester.

Just, the thing I am trying to achieve, is to have the ability to persist the Pestr Object state on a file system for future reuse. 
Currently, the only way to persist a Pester run on disk, is to export it to NunitXML. 
Convertir an XML to a JSON format using `ConvertTo-JSON` is a pain. 
Plus, it is not 100% identical to the PesterObject.

It would make everything much much easier if JSON was supported out of the box in invoke-pester, as that format is easier to use. It will allow external frameworks to be able to work on that format and easily extend some of the Pesterfunctionality (With reporting for example.... I have private project coming doing just that using PSHTML. But JSON makes everything SO much simpler).

**Comment by nohwnd** [comment]
Then I am probably confused about what you want to do. 

Is this, pretty much what you want to have, but inside of pester so you can tell it to save as json and give it the path? 

```powershell
$object = invoke-pester C:\folder\Tests -PassThru
$object | ConvertTo-Json | out-File -FilePath .\TestResults.json 
```

Or the functionality will be more involved?

If yes, why would you need to convert to xml and then from xml to json when you can save directly to json?

**Comment by Stephanevg** [comment]
Yeah, that is actually what I would like to have from `invoke-pester` directly.

No, my point with the XML was to say that currently the only way to persist the `Pester data` on file, is the NunitXML format. 
I used that to persist Infrastructure tests in my environement.
I realized that what I actually wanted, is the Pesterobject, not the NunitXML, as it is more complex to work with. 

So, saving the object to a JSON file would really make things easier for parsing past Pester runs for example. It is a small change, but I think it will beneficial to a lot of scripters.
(In my team at work, and from the French Powershell community, there have been demands for this feature).

> Looking at the PR it seems like a simple change so why not. :)

I have the done the modifications as wished. Is the implementation ok for you @nohwnd ? can I go ahead and write the tests? ;)

**Comment by nohwnd** [comment]
@Stephanevg yes please write the tests.

**Comment by Stephanevg** [comment]
Hi @nohwnd 
I am coming back to this one after seeing the PR from BGelens, it remminded me I also had a feature pending here.

As I was mentionning in my PR https://github.com/pester/Pester/pull/1338 Convertto-Json is only supported from Powershell v3. Therefor, we are a bit 'blocked' to add it into Pester 4.X.

I could write a 'custom' JSON converter, but I am not so sure it will make sense, as it might just be some additional code to maintain.

I remeber you were talking at PSConfeu that you would like to shift the support from 2.0 to 3.0 in Pester version 5. This feature would confort that that shift is actually needed.

What do you think we should do about it?

**Comment by Stephanevg** [comment]
Hi @nohwnd 
You are probably super buys or so. Could you just tel me if it would be possible to add this to v 5.0 (And confirm that the dependency on Powershell 2.0 will not be present in v5.0 anymore).

Otherwise, we can always try to build our own json parser but that is something I would defenitly not suggest.

**Comment by renehernandez** [comment]
@Stephanevg Yes, Pester v5.0 only supports PowerShell 3+.

You can close the current pull request and then open a new one against v5.0. You may have to do a little bit of rewrite to adapt to the new structure of version 5.

Let us know if you need any help

**Comment by nohwnd** [comment]
@Stephanevg I changed the base of your PR to branch v5.0, I would recommend deleting it locally and then fetching the branch from the server again. 

https://github.com/pester/Pester/pull/1338

**Comment by Stephanevg** [comment]
Cool, thanks for doing that @nohwnd I'll have a look at that as soon as I have a moment.

I was looking into some of the JSON data I got as output, and noticed that there is no way to garantee that the JSON output (or the PesterObject itself) is of type 'Pester'.

Let me explain:

I am rereading JSON exported PesterObjects that have been persisted on a filesystem. I have no trusthworthfull way to identifiy that that json object IS really a PesterObject.

for nunitXML we have the following node with ```Name=pester``` which can guarantee that.

```
<test-results xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="nunit_schema_2.5.xsd" name="Pester" total="711" errors="0" failures="0" not-run="0" inconclusive="0" ignored="0" skipped="0" invalid="0" date="2019-03-27" time="22:11:50">
```

I think that the JSON export should also have something that goes along these lines.

## Implementation suggestion

I would suggest we extend the pester object with a property called 'Name' or 'Type' and set it to Pester.
Like that, we have the information available per code when using ``` -Paththru``` and also when we will export it to JSON.

If this is ok, I can immediatley add it to my PR. Let me know what you guys think!

**Comment by Stephanevg** [comment]
## Suggestion 2
An alternate (or additional?) solution, would be to export the file using a specific file extension, like

-  ```.Pester``` or  
- ``` .Po``` for ```PesterObject``` 
- or ```.ppo``` for ``` PowershellPesterObject ```

Anyone has a better one perhaps ?

> This probably needs to have some kind of discussion first

**Comment by nohwnd** [comment]
Neither of those will ensure that you will get a properly formatted json file that contains Pester data. 
1) the way to do this is by JSON schema. That is the way to define type information and structure of JSON documents. But it might be a bit difficult for this schema, especially since it is recursive.
2) I would not want to rely on names figure out the content of a file, similarly as `.Tests.ps1` is helpful, but in no way required to identify a test file. 

Looking back at this I am still not sure what you are trying to achieve here, and how it is beneficial for v5 with the described complications. The result object is itself structured to contain the raw data that can be stored for further processing, and that can be done by any means. Using `ConvertTo-Json` or `Export-CliXml` or some other way. The user can then use whatever naming, and infrastructure to save a test result. 

So is there even a need to have a schema or identify the export files? Because this will still be compatible with just a "single" version of Pester, because making sure the format is compatible among versions is complicated, and very likely not worth it. 

So all in all the user has to know what they are doing, and when they say the result files are Pester result files, they better be Pester result files.

**Comment by Stephanevg** [comment]
Yeah you are right @nohwnd , then it might make sense just to focus on an extension, without necessarily defining a schema?

The extension ```.Tests.ps1``` became a standard in the Powershell community, and I have rarely seen people writing their tests without that extension in it. It helps identify the file type containing pester tests (without guaranteeing it).

Adding a specific extension won't guarantee that the file actually contains the data we expect - but - people could rely on it, just as they do for the ```.tests.ps1``` one.

As ```.Json``` is to generic in my opinion, having a possibility to sort it out on something specific to pester (like ```.Pester.Json``` / ```.pester```  / ```.po```... ) would really make things easier to at least ```give a hint``` that that file is a pester export.

I don't think it will cost anything in terms of maintainability and it will add some structure to the export.

**Comment by nohwnd** [comment]
I think we agreed to standardize the way to serialize the run object to clixml not JSON.
